### ISSUE TITLE
RyuJIT/x86: CoreFx System.Linq.Tests failure

### ISSUE THREAD

**Issue by RussKeldorph** [dotnet/coreclr#8163]
**Repro steps:**

coreclr *(assuming root=c:\git\coreclr)*
```
build x86 checked skiptests
```
corefx *(assuming root=c:\git\corefx)*
```
build -Release -SkipTests -TestNugetRuntimeId=win7-x86 -- /p:BUILDTOOLS_OVERRIDE_RUNTIME=c:\git\coreclr\bin\Product\Windows_NT.x86.Checked
cd bin\tests\AnyOS.AnyCPU.Release\System.Linq.Tests\default.netcoreapp1.1
set COMPlus_AltJit=*
set COMPlus_AltJitName=protojit.dll
RunTests.cmd c:\git\corefx\packages
```
**Output**
```
Using c:\git\corefx\packages as folder for resolving package dependencies.
Executing in C:\git\corefx\bin\tests\AnyOS.AnyCPU.Release\System.Linq.Tests\default.netcoreapp1.1\
Hard linking dependent files...
Finished linking needed files, moving to running tests.
Running tests... Start time: 17:07:24.82
Command(s):
call CoreRun.exe xunit.console.netcore.exe System.Linq.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.1tests  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Linq.Tests
Discovered:  System.Linq.Tests
Starting:    System.Linq.Tests
   System.Linq.Tests.MaxTests.Max_NullableLong(source: [null, -16, 0, 50, 100, ...], expected: 1000) [FAIL]
      Assert.Equal() Failure
      Expected: 1000
      Actual:   -16
      Stack Trace:
            at System.Linq.Tests.MaxTests.Max_NullableLong(IEnumerable`1 source, Nullable`1 expected)
   System.Linq.Tests.MaxTests.Max_NullableLong(source: [null, -16, 0, 50, 100, ...], expected: 9223372036854775807) [FAIL]
      Assert.Equal() Failure
      Expected: 9223372036854775807
      Actual:   -16
      Stack Trace:
            at System.Linq.Tests.MaxTests.Max_NullableLong(IEnumerable`1 source, Nullable`1 expected)
   System.Linq.Tests.MaxTests.Max_NullableLong(source: [-6, null, null, 0, -9, ...], expected: 0) [FAIL]
      Assert.Equal() Failure
      Expected: 0
      Actual:   -6
      Stack Trace:
            at System.Linq.Tests.MaxTests.Max_NullableLong(IEnumerable`1 source, Nullable`1 expected)
   System.Linq.Tests.MaxTests.Max_NullableLong_WithSelectorAccessingField [FAIL]
      Assert.Equal() Failure
      Expected: 9223372036854775807
      Actual:   -105
      Stack Trace:
            at System.Linq.Tests.MaxTests.Max_NullableLong_WithSelectorAccessingField()
Finished:    System.Linq.Tests

=== TEST EXECUTION SUMMARY ===
   System.Linq.Tests  Total: 2190, Errors: 0, Failed: 4, Skipped: 0, Time: 71.678s
Finished running tests.  End time=17:08:53.81, Exit code = 1
```

**Comment by mikedn** [comment]
Looks like `CodeGen::genCompareLong` is broken. LINQ's `Max` has this code:
```C#
while (e.MoveNext())
{
    long? cur = selector(e.Current);
    long x = cur.GetValueOrDefault();
    if (cur.HasValue & x > valueVal)
    {
        valueVal = x;
        value = cur;
    }
}
```
and the JIT generates:
```
       8D55B0       lea      edx, bword ptr [ebp-50H]
       8BC6         mov      eax, esi
       8B4804       mov      ecx, gword ptr [eax+4]
       FF500C       call     dword ptr [eax+12]System.Func`2[Nullable`1,Nullable`1][System.Nullable`1[System.Int64],System.Nullable`1[System.Int64]]:Invoke(struct):struct:this
       8B4DB8       mov      ecx, dword ptr [ebp-48H]
       8B45BC       mov      eax, dword ptr [ebp-44H]
       3BC3         cmp      eax, ebx
       7510         jne      SHORT G_M55891_IG13
       3BCF         cmp      ecx, edi
       0F97C2       seta     dl
       0FB6D2       movzx    edx, dl
       899568FFFFFF mov      dword ptr [ebp-98H], edx
       EB06         jmp      SHORT G_M55891_IG14

G_M55891_IG13:
       0F9FC2       setg     dl
       0FB6D2       movzx    edx, dl

G_M55891_IG14:
       8B55B0       mov      edx, dword ptr [ebp-50H]
       0FB6D2       movzx    edx, dl
       239568FFFFFF and      edx, dword ptr [ebp-98H]
       7417         je       SHORT G_M55891_IG15
```
There's a long compare that produces a result and it spills it to `[ebp-98h]`. But it does that only on one branch, in IG13 `edx` is not spilled and then it is killed at the start of IG14.

That's probably because codegen calls `genProduceReg` twice:
```C++
        inst_SET(jumpKindLo, targetReg);
        // Set the higher bytes to 0
        inst_RV_RV(ins_Move_Extend(TYP_UBYTE, true), targetReg, targetReg, TYP_UBYTE, emitTypeSize(TYP_UBYTE));
        genProduceReg(tree); // SHOULD NOT BE HERE!

        inst_JMP(EJ_jmp, labelFinal);

        // Define the label for hi jump target here. If we have jumped here, we want to set
        // the target register based on the jump kind of the actual compare type.

        genDefineTempLabel(labelHi);
        inst_SET(genJumpKindForOper(tree->gtOper, compareKind), targetReg);

        // Set the higher bytes to 0
        inst_RV_RV(ins_Move_Extend(TYP_UBYTE, true), targetReg, targetReg, TYP_UBYTE, emitTypeSize(TYP_UBYTE));
        genProduceReg(tree);
```

I'll try to make a fix.

**Comment by mikedn** [comment]
Yep, all LINQ tests pass with the fix in #8182.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib
